src/main/java/gregtech/api/capability/impl/FuelRecipeLogic.java

```diff
@@ -207,21 +224,88 @@ public long getMaxVoltage() {
         return maxVoltage;
     }
 
+    protected double calculateFuelConsumptionMultiplier() {
+        return 1.0;
+    }
+
     protected int calculateFuelAmount(FuelRecipe currentRecipe) {
-        return currentRecipe.getRecipeFluid().amount * getVoltageMultiplier(getMaxVoltage(), currentRecipe.getMinVoltage());
+        return (int) (currentRecipe.getRecipeFluid().amount * getVoltageMultiplier(getMaxVoltage(), currentRecipe.getMinVoltage()) * calculateFuelConsumptionMultiplier());
+    }
+
+    public int getFuelAmount() {
+        return calculateFuelAmount(this.previousRecipe);
+    }
+
+    protected double calculateRecipeDurationMultiplier() {
+        return 1.0;
     }
 
     protected int calculateRecipeDuration(FuelRecipe currentRecipe) {
-        return currentRecipe.getDuration();
+        return (int) (currentRecipe.getDuration() * calculateRecipeDurationMultiplier());
+    }
+
+    public int getRecipeDuration() {
+        return calculateRecipeDuration(this.previousRecipe);
+    }
+
+    /**
+     * this method aim to tell if the current recipe has ended in order to check for a new recipe see {@link #isActive}
+     * or {@link #isWorkingEnabled} if you want to know if the machine is running.
+     *
+     * @return true if no recipe is running false otherwise
+     */
+    public boolean hasRecipeEnded() {
+        return this.recipeDurationLeft <= 0;
+    }
+
+    public boolean canProduceEnergy() {
+        return !hasRecipeEnded();
+    }
+
+    public boolean canConsumeFuel() {
+        return !hasRecipeEnded();
+    }
+
+    /**
+     * Compute a multiplier used by default in {@link #startRecipe(FuelRecipe, int, int)}  startRecipe} each time the recipe start.
+     * fast changing efficiency could be implemented in {@link #calculateDynamicEnergyEfficiency() calculateDynamicEnergyEfficiency} for better responsiveness of the changes.
+     *
+     * @return power multiplier
+     */
+    protected double calculateStaticEnergyEfficiency() {
+        return 1.0;
+    }
+
+    /**
+     * Compute a multiplier used default in {@link #calculateRecipeOutputVoltage() calculateRecipeOutputVoltage} each tick.
+     * constant/rarely changing efficiency should/could be implemented in {@link #calculateStaticEnergyEfficiency() calculateStaticEnergyEfficiency}
+     *
+     * @return power multiplier
+     */
+
+    protected double calculateDynamicEnergyEfficiency() {
+        return 1.0;
+    }
+
+    /**
+     * Apply {@link #calculateDynamicEnergyEfficiency() calculateDynamicEnergyEfficiency} multiplier
+     * to the outputVoltage returned by {@link #getRecipeOutputVoltage(), getRecipeOutputVoltage}
+     * in order to get the effective power output used in the {@link #update() update} method.
+     *
+     * @return Effective power output of the machine.
+     */
+    public long calculateRecipeOutputVoltage() {
```

Comment:
If you can override this method, why do you even have calculateDynamicEnergyEfficiency?
Again, what prevents you from just doing super.calculateRecipeOutputVoltage() * whateverYouWant in your child class?

Reply:
how about `calculateFuelAmount` and `calculateRecipeDuration` and their respective multiplier ?
for me this is more about make things clear, what methods does what. Each things does one thing and do it clearly and properly. 

`calculateRecipeOutputVoltage` get all the different multipliers and apply them in his way but doesn't touch to how are those multipliers are computed.

Reply:
I understand your desire to keep things clean, but you just end up making them more bloated than they should be.
I really think in that particular case it's better to have very few methods for changing functionality, granted they're all flexible and simple to use and understand.

Reply:
Like, I would literally just leave 3 methods for overrides: calculateRecipeDuration, calculateRecipeOutputVoltage and calculateFuelAmount, plus have their respective getters as final methods. That way it's very minimalistic and simple to understand.
Obviously that does not cover all functionality and does not allow overriding fuel consumption/energy production conditions, so that's why we introduce canProduceEnergy and shouldConsumeFuel, which ideally now contain logic that used to be hard-coded inside of the update() method
Plus it's probably a good idea to keep some of the utility methods for customizing default functionality without fully rewriting it, like shouldVoidExcessiveEnergy (or whatever that method was called)

That's it, I would literally keep only these few methods and nothing else

Reply:
I think that the energy voiding policy should be placed in an other method to keep things splitted this way warjot will be happy because those two policy will be in different methods and won't be hard coded what do think of that ? 

considering reducing the number of methods I think both ways are acceptable I prefer to have different methods to keep things splitted . I will ask others council member their opinion concerning this question.

Reply:
as Dstrand1 agree with you I will apply your changes